File: packages/core/src/results.ts
import { mkdir, readFile, writeFile } from "node:fs/promises";
import { dirname, join } from "node:path";

import { Effect, Schema } from "effect";
import { parse, stringify } from "yaml";

import {
  TestResultsFileReadError,
  type TestResultsLoadError,
  TestResultsSchemaDecodeError,
  TestResultsYamlParseError,
} from "./errors.ts";
import type { PromiseRunStatus } from "./schema.ts";

export const harnessResultsPath = ".harness/results.yaml";
export const harnessRootEnvVar = "HARNESS_ROOT_DIR";

export const TestResultStatusSchema = Schema.Literals(["passing", "failing", "skipped"]);

export const TestResultSchema = Schema.Struct({
  failureMessage: Schema.optionalKey(Schema.String),
  file: Schema.String,
  promiseId: Schema.String,
  status: TestResultStatusSchema,
  testName: Schema.String,
});

export const TestResultsFileSchema = Schema.Struct({
  generatedAt: Schema.String,
  results: Schema.Array(TestResultSchema),
});

export type TestResult = Schema.Schema.Type<typeof TestResultSchema>;
export type TestResultsFile = Schema.Schema.Type<typeof TestResultsFileSchema>;
export type TestResultStatus = Schema.Schema.Type<typeof TestResultStatusSchema>;

const resolveResultsPath = (rootDir: string): string => join(rootDir, harnessResultsPath);

const readResultsFile = (path: string): Effect.Effect<string, TestResultsFileReadError> =>
  Effect.tryPromise({
    try: () => readFile(path, "utf8"),
    catch: (cause) => new TestResultsFileReadError({ cause, path }),
  });

const parseResultsYaml = (
  path: string,
  raw: string,
): Effect.Effect<unknown, TestResultsYamlParseError> =>
  Effect.try({
    try: () => parse(raw),
    catch: (cause) => new TestResultsYamlParseError({ cause, path }),
  });

const decodeResultsFile = (
  path: string,
  input: unknown,
): Effect.Effect<TestResultsFile, TestResultsSchemaDecodeError> =>
  Schema.decodeUnknownEffect(TestResultsFileSchema)(input).pipe(
    Effect.mapError((cause) => new TestResultsSchemaDecodeError({ cause, path })),
  );

export const loadTestResultsFile = (
  rootDir: string,
): Effect.Effect<TestResultsFile | undefined, TestResultsLoadError> => {
  const path = resolveResultsPath(rootDir);
  return Effect.gen(function* () {
    const raw = yield* readResultsFile(path).pipe(
      Effect.catchTag("TestResultsFileReadError", (error) => {
        const nodeError = error.cause as { readonly code?: string };
        if (nodeError.code === "ENOENT") return Effect.succeed(undefined);
        return Effect.fail(error);
      }),
    );
    if (raw === undefined) return undefined;
    const parsed = yield* parseResultsYaml(path, raw);
    return yield* decodeResultsFile(path, parsed);
  });
};

export const loadTestResults = (
  rootDir: string,
): Effect.Effect<readonly TestResult[], TestResultsLoadError> =>
  loadTestResultsFile(rootDir).pipe(Effect.map((file) => file?.results ?? []));

export const writeTestResultsFile = async (
  rootDir: string,
  file: TestResultsFile,
): Promise<void> => {
  const path = resolveResultsPath(rootDir);
  await mkdir(dirname(path), { recursive: true });
  await writeFile(
    path,
    stringify(file, {
      lineWidth: 0,
    }),
  );
};

export const createTestResultsFile = (
  results: readonly TestResult[],
  generatedAt = new Date().toISOString(),
): TestResultsFile => ({
  generatedAt,
  results: [...results].sort((left, right) => {
    const byPromise = left.promiseId.localeCompare(right.promiseId);
    if (byPromise !== 0) return byPromise;
    const byFile = left.file.localeCompare(right.file);
    if (byFile !== 0) return byFile;
    return left.testName.localeCompare(right.testName);
  }),
});

export const getPromiseRunStatus = (
  promiseId: string,
  results: readonly TestResult[],
): PromiseRunStatus => {
  const matched = results.filter((result) => result.promiseId === promiseId);
  if (matched.length === 0) return "unknown";
  if (matched.some((result) => result.status === "failing")) return "failing";
  if (matched.some((result) => result.status === "skipped")) return "skipped";
  return "passing";
};

File: crates/harness-core/src/results.rs
use crate::errors::HarnessError;
use harness_protocol::{
    decode_test_results_file, PromiseRunStatus, ProtocolDecodeError, ProtocolVersion, TestResult,
    TestResultStatus, TestResultsFile,
};
use std::fs;
use std::path::Path;

pub const HARNESS_RESULTS_PATH: &str = ".harness/results.yaml";
pub const HARNESS_ROOT_ENV_VAR: &str = "HARNESS_ROOT_DIR";

pub fn load_test_results_file(
    root_dir: impl AsRef<Path>,
) -> Result<Option<TestResultsFile>, HarnessError> {
    let path = root_dir.as_ref().join(HARNESS_RESULTS_PATH);
    let raw = match fs::read_to_string(&path) {
        Ok(raw) => raw,
        Err(error) if error.kind() == std::io::ErrorKind::NotFound => return Ok(None),
        Err(error) => {
            return Err(HarnessError::TestResultsFileReadError {
                path,
                cause: error.to_string(),
            })
        }
    };
    decode_test_results_file(&raw)
        .map_err(|error| match error {
            ProtocolDecodeError::Yaml(cause) => HarnessError::TestResultsYamlParseError {
                path: path.clone(),
                cause,
            },
            ProtocolDecodeError::Shape(cause) => HarnessError::TestResultsSchemaDecodeError {
                path: path.clone(),
                cause,
            },
        })
        .map(Some)
}

pub fn load_test_results(root_dir: impl AsRef<Path>) -> Result<Vec<TestResult>, HarnessError> {
    Ok(load_test_results_file(root_dir)?.map_or_else(Vec::new, |file| file.results))
}

pub fn create_test_results_file(
    mut results: Vec<TestResult>,
    generated_at: impl Into<String>,
) -> TestResultsFile {
    results.sort_by(|left, right| {
        left.promise_id
            .cmp(&right.promise_id)
            .then_with(|| left.file.cmp(&right.file))
            .then_with(|| left.test_name.cmp(&right.test_name))
    });
    TestResultsFile {
        api_version: ProtocolVersion,
        generated_at: generated_at.into(),
        results,
    }
}

pub fn write_test_results_file(
    root_dir: impl AsRef<Path>,
    file: &TestResultsFile,
) -> Result<(), HarnessError> {
    let path = root_dir.as_ref().join(HARNESS_RESULTS_PATH);
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent).map_err(|error| {
            HarnessError::TestResultsDirectoryCreateError {
                path: parent.to_path_buf(),
                cause: error.to_string(),
            }
        })?;
    }
    let raw = serde_yaml::to_string(file).map_err(|error| {
        HarnessError::TestResultsSchemaDecodeError {
            path: path.clone(),
            cause: error.to_string(),
        }
    })?;
    fs::write(&path, raw).map_err(|error| HarnessError::TestResultsFileWriteError {
        path,
        cause: error.to_string(),
    })
}

pub fn get_promise_run_status(promise_id: &str, results: &[TestResult]) -> PromiseRunStatus {
    let matched = results
        .iter()
        .filter(|result| result.promise_id == promise_id)
        .collect::<Vec<_>>();
    if matched.is_empty() {
        return PromiseRunStatus::Unknown;
    }
    if matched
        .iter()
        .any(|result| result.status == TestResultStatus::Failing)
    {
        return PromiseRunStatus::Failing;
    }
    if matched
        .iter()
        .any(|result| result.status == TestResultStatus::Skipped)
    {
        return PromiseRunStatus::Skipped;
    }
    PromiseRunStatus::Passing
}
